CMakeLists.txt

```diff
@@ -0,0 +1,119 @@
+cmake_minimum_required(VERSION 2.8.12)
+
+project(cctz)
+
+set(CMAKE_MODULE_PATH
+  ${PROJECT_SOURCE_DIR}/cmake
+  ${PROJECT_SOURCE_DIR}/cmake/modules
+  ${CMAKE_MODULE_PATH})
+
+include(CheckCXXSymbolExists)
+include(CTest)
+
+option(BUILD_EXAMPLES "Whether or not to build examples" ON)
+
+function(cctz_target_set_cxx_standard target)
+  set(cctz_cxx_standard 11)
+  # CXX_STANDARD target property is only supported since CMake >= 3.1
+  if (CMAKE_VERSION VERSION_LESS "3.1")
+    if (CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+      target_compile_options(${target} PRIVATE -std=c++${cctz_cxx_standard})
+    else()
+      # if a specific standard is required, this can be set from the command line:
+      #     cmake -DCMAKE_CXX_STANDARD=[11|14|17]
+      set_property(TARGET ${target} PROPERTY CXX_STANDARD ${cctz_cxx_standard})
+    endif()
+  endif()
+endfunction()
+
+add_library(cctz
+  src/civil_time_detail.cc
+  src/time_zone_fixed.cc
+  src/time_zone_fixed.h
+  src/time_zone_format.cc
+  src/time_zone_if.cc
+  src/time_zone_if.h
+  src/time_zone_impl.cc
+  src/time_zone_impl.h
+  src/time_zone_info.cc
+  src/time_zone_info.h
+  src/time_zone_libc.cc
+  src/time_zone_libc.h
+  src/time_zone_lookup.cc
+  src/time_zone_posix.cc
+  src/time_zone_posix.h
+  src/tzfile.h
+  src/zone_info_source.cc
+  )
+cctz_target_set_cxx_standard(cctz)
+add_library(cctz::cctz ALIAS cctz)
+
+target_include_directories(cctz PUBLIC include)
+
+# the getopt_long function used by time_tool is not available on all platforms,
+# for example this is not available by default on Windows
+if (NOT DEFINED HAVE_GETOPT_LONG)
+  set(show_time_tool_msg 1)
+endif()
+check_cxx_symbol_exists(getopt_long getopt.h HAVE_GETOPT_LONG)
+if (HAVE_GETOPT_LONG)
+  add_executable(time_tool src/time_tool.cc)
+  cctz_target_set_cxx_standard(time_tool)
+  target_link_libraries(time_tool cctz::cctz)
+elseif(show_time_tool_msg)
+    message(STATUS "Disable time_tool as getopt_long is not available")
+endif()
+
+if (BUILD_EXAMPLES)
```

Comment:
Can BUILD_EXAMPLES be CCTZ_BUILD_EXAMPLES and BUILD_TESTING be CCTZ_BUILD_TESTING?

My project pulls cctz in via add_subdirectory() (to avoid the external projects inheritance mess) and I would hate to have to set something so generic sounding one level up.

Reply:
I would prefer not to.
I think this is the expected, simplest, way to use options.
`BUILD_TESTING` and other options like `BUILD_SHARED_LIBS`, are actually declared by CMake.
I'm seeing quite a few project namespacing these options to specifically for CMakeLists.txt that uses `add_subdirectory(cctz)`.
But I don't think `add_subdirectory()` needs any special treatment, I have actually realized this when someone on the Abseil CMake PR made me realize how safe `add_subdirectory` was.

About a week ago I think, I saw someone talk about this exact problem on the cppclang Slack, in the \#cmake channel, and the exposed solution is both simple and clean IMHO.

The super-project can just wrap the `add_subdirectory` call with the appropriate value for `BUILD_TESTING` and friends.
It can decide to:
- disable it or enable it
- to have a specific option `CCTZ_BUILD_EXAMPLES` 
- or a grouped option like `DEPS_BUILD_EXAMPLES`
- do anything else, even nothing

I haven't actually tested, so I hope this is correct...let me know.